### PR TITLE
[FIXED JENKINS-41668] Add "dir" option for Dockerfile.

### PR DESCRIPTION
**Pull request by abayer**
* JENKINS issue(s):
    * [JENKINS-41668](https://issues.jenkins-ci.org/browse/JENKINS-41668)
* Description:
    * In order to build a Dockerfile in a different directory, we need to also specify the directory, not just the filename, so let's add a `dir` option to the `dockerfile` declarative agent.
* Documentation changes:
    * https://github.com/jenkins-infra/jenkins.io/pull/588
* Users/aliases to notify:
    * @reviewbybees 
    * @rsandell

### CODE REVIEW

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy [hunk 1]
```diff
@@ -588,6 +588,17 @@ class ModelValidatorImpl implements ModelValidator {
                         }
                     }
                     map.variables.each { k, v ->
+                        // TODO: JENKINS-41746 - get rid of this hardcoding and use extensible validation instead once we can
+                        if (typeName == "dockerfile") {
+                            if ((k.key == "dir" || k.key == "filename") && v.toGroovy().contains(" ")) {
```
Comment (rsandell) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy]: check for regexp '\s+' instead, to cover tabs and new lines as well.

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy [hunk 2]
```diff
@@ -588,6 +588,17 @@ class ModelValidatorImpl implements ModelValidator {
...
+                                errorCollector.error(k, Messages.ModelValidatorImpl_SpaceInDockerfilePath(k.key))
+                                valid = false
+                            }
+                            if (k.key == "filename" && (v.toGroovy().contains("/") || v.toGroovy().contains("\\"))) {
+                                errorCollector.error(k, Messages.ModelValidatorImpl_DirSeparatorInDockerfileName())
```
Comment (rsandell) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy]: wasn't this possible to do in 1.0? and are we breaking behavior by not allowing this?

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/DockerPipelineFromDockerfile.java [hunk 3]
```diff
@@ -52,6 +53,23 @@ public void setFilename(String filename) {
...
+        return dir;
+    }
+
+    @DataBoundSetter
+    public void setDir(String dir) {
+        this.dir = dir;
+    }
+
+    public String getActualDir() {
+        if (dir != null) {
```
Comment (rsandell) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/DockerPipelineFromDockerfile.java]: 'StringUtils.isEmpty()'

File: pipeline-model-definition/src/test/resources/errors/spaceInDockerfileDir.groovy [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+pipeline {
+    agent {
+        dockerfile {
+            args "-v /tmp:/tmp -p 8000:8000"
+            dir "sub dir"
```
Comment (rsandell) [inline pipeline-model-definition/src/test/resources/errors/spaceInDockerfileDir.groovy]: what about 'sub\tdir'?

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy (hunk 1 above)
Comment (abayer) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy]: Changed my mind - gonna do better quoting instead. =)

File: pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy (hunk 2 above)
Comment (abayer) [inline pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy]: So it was possible to do in 1.0 but it broke, hence JENKINS-41668 in the first place. =)

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/DockerPipelineFromDockerfile.java (hunk 3 above)
Comment (abayer) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/DockerPipelineFromDockerfile.java]: *nod*

File: pipeline-model-definition/src/test/resources/errors/spaceInDockerfileDir.groovy (hunk 4 above)
Comment (abayer) [inline pipeline-model-definition/src/test/resources/errors/spaceInDockerfileDir.groovy]: As mentioned above, I'm junking the space check and just doing better quoting.

File: pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/DockerPipelineFromDockerfile.java [hunk 5]
```diff
@@ -52,6 +55,33 @@ public void setFilename(String filename) {
         this.filename = filename;
     }
 
+    public @Nonnull String getDir() {
```
Comment (rsandell) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/DockerPipelineFromDockerfile.java]: This is a lie; `dir` will always be `null` until `setDir` has been called, and even then it allows `null` values to be set.

I guess you intended to put `@Nonnull` on `getActualDir`?

Comment (abayer) [inline pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/impl/DockerPipelineFromDockerfile.java]: Whoops - fixing.

**Comment by crd477** [comment]
I have a pipeline that uses:

`agent { dockerfile { dir 'script/docker/ubuntu-xenial-amd64' } }`

The Dockerfile in that directory works exactly as I want when manually invoking `docker build` from the command line.  Yet, when the pipeline executes, Jenkins attempts:

`docker build -t b70eaf9d3cae6cbc7c236da74233474192424290 -f script/docker/ubuntu-xenial-amd64/Dockerfile script/docker/ubuntu-xenial-amd64`

What I really want it to execute is:

`docker build -t b70eaf9d3cae6cbc7c236da74233474192424290 -f script/docker/ubuntu-xenial-amd64/Dockerfile .`

(Note the "." at the end).  How can I accomplish that?

**Comment by sschuberth** [comment]
Very good question, @crd477. It seems the implementation by @abayer always assumes `filename` to be relative to `dir`, i.e. it does not allow to set the context directory completely independent of the *Dockerfile*, which is a feature I'm looking for, too.

**Pull request merged**